### ISSUE TITLE
ESBMC fails to verify case with pointer casting

### ISSUE THREAD

**Issue by feliperodri** [esbmc/esbmc#27]
ESBMC fails to verify the following case:

```
typedef unsigned long size_t;
typedef unsigned char uint8_t;
typedef unsigned long int uintptr_t;

struct store_byte_from_buffer {
    unsigned long index;
    unsigned char byte;
};

struct aws_byte_cursor {
    unsigned long len;
    unsigned char *ptr;
};

_Bool nondet_bool() {
    _Bool nondet;
    return nondet;
}

size_t nondet_size_t() {
    size_t nondet;
    return nondet;
}

void *bounded_malloc(size_t size) {
    __ESBMC_assume(size <= ((18446744073709551615UL) >> (8 + 1)));
    return malloc(size);
}

_Bool aws_byte_cursor_is_valid(const struct aws_byte_cursor *cursor) {
    return cursor &&
           ((cursor->len == 0) || (cursor->len > 0 && cursor->ptr && ((((cursor->len)) == 0) || ((cursor->ptr)))));
}

void assert_bytes_match(const uint8_t *const a, const uint8_t *const b, const size_t len) {
    __ESBMC_assert(!a == !b, "This should never fail");
    if (len > 0 && a != ((void *)0) && b != ((void *)0)) {
        size_t i;
        __ESBMC_assume(i < len && len < ((18446744073709551615UL) >> (8 + 1)));
        __ESBMC_assert(a[i] == b[i], "This should never fail");
    }
}

size_t aws_nospec_mask(size_t index, size_t bound) {
    __asm__ __volatile__("" : "+r"(index));
    size_t negative_mask = index | bound;
    size_t toobig_mask = bound - index - (uintptr_t)1;
    size_t combined_mask = negative_mask | toobig_mask;
    combined_mask = (~combined_mask) / (
                                       (18446744073709551615UL)
                                                - (
                                                   (18446744073709551615UL)
                                                            >> 1));







    combined_mask = combined_mask *
                                   (18446744073709551615UL)
                                              ;

    return combined_mask;
}

struct aws_byte_cursor aws_byte_cursor_advance_nospec(struct aws_byte_cursor *const cursor, size_t len) {
    __ESBMC_assume((aws_byte_cursor_is_valid(cursor)));

    struct aws_byte_cursor rv;

    if (len <= cursor->len && len <= (
                                     (18446744073709551615UL)
                                              >> 1) && cursor->len <= (
                                                                       (18446744073709551615UL)
                                                                                >> 1)) {





        uintptr_t mask = aws_nospec_mask(len, cursor->len + 1);


        len = len & mask;
        cursor->ptr = (uint8_t *)((uintptr_t)cursor->ptr & mask);

        cursor->len = cursor->len & mask;

        rv.ptr = cursor->ptr;

        rv.len = len & mask;

        cursor->ptr += len;
        cursor->len -= len;
    } else {
        rv.ptr =
                ((void *)0)
                    ;
        rv.len = 0;
    }

    __ESBMC_assert((aws_byte_cursor_is_valid(cursor)), "This should never fail");
    __ESBMC_assert((aws_byte_cursor_is_valid(&rv)), "This should never fail");
    return rv;
}

_Bool aws_byte_cursor_read(struct aws_byte_cursor *restrict cur, void *restrict dest, const size_t len) {
    __ESBMC_assume((aws_byte_cursor_is_valid(cur)));
    __ESBMC_assume((((((len)) == 0) || ((dest)))));
    struct aws_byte_cursor slice = aws_byte_cursor_advance_nospec(cur, len);

    if (slice.ptr) {
        memcpy(dest, slice.ptr, len);
        __ESBMC_assert((aws_byte_cursor_is_valid(cur)), "This should never fail");
        __ESBMC_assert((((((len)) == 0) || ((dest)))), "This should never fail");
        return 1;
    }
    __ESBMC_assert((aws_byte_cursor_is_valid(cur)), "This should never fail");
    return 0;
}

int main() {

    struct aws_byte_cursor cur;
    size_t length;
    void *dest = bounded_malloc(length);
    
    cur.ptr = bounded_malloc(cur.len);
    
    __ESBMC_assume(aws_byte_cursor_is_valid(&cur));


    __ESBMC_assume(((((length)) == 0) || ((dest))));


    struct aws_byte_cursor old_cur = cur;


    if (aws_byte_cursor_read(&cur, dest, length) == 1) {
        __ESBMC_assert(!old_cur.ptr == !dest, "This should never fail");
        if (length > 0 && old_cur.ptr != ((void *)0) && dest != ((void *)0)) {
            size_t i;
            __ESBMC_assume(i < length && length < ((18446744073709551615UL) >> (8 + 1)));
            __ESBMC_assert(old_cur.ptr[i] == ((uint8_t *)dest)[i], "This should never fail");
        }
    }
}

```

Command:
```
esbmc --no-div-by-zero-check --force-malloc-success --state-hashing --no-align-check --k-step 2 --floatbv --unlimited-k-steps --context-bound 2 --64 --k-induction --max-inductive-step 3 --no-pointer-check --no-bounds-check --interval-analysis main.c
```

I also compared with CBMC using the following command and it passes:
```
cbmc --bounds-check --div-by-zero-check --float-overflow-check --nan-check --pointer-check --pointer-overflow-check --signed-overflow-check --undefined-shift-check --unsigned-overflow-check --unwind 4 --unwinding-assertions  --object-bits 8 --trace main.c -D__ESBMC_assume=__CPROVER_assume -D__ESBMC_assert=__CPROVER_assert
```

**Comment by mikhailramalho** [comment]
Commit 2055c0ebd3 seems to fix the issue.

I compared our get_value_set_rec with CBMC's one and there are only a couple of things missing:
1. Support for the bit operations (addressed by cce5c75).
2. Support for mult. CBMC gives the following example: `(int*)((sel*(ulong)&a)+((sel^0x1)*(ulong)&b))`
3. Support for byte_update. The code seems similar to the byte_extract one.

**Comment by mikhailramalho** [comment]
Fixed by 2055c0e .
